modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/rule/AbstractIgniteJoinConverterRule.java

```diff
@@ -0,0 +1,192 @@
+    /** Known join type hints and the opposite hints. */
+    private static final EnumMap<HintDefinition, HintDefinition> HINTS = new EnumMap<>(HintDefinition.class);
+
+    /** Hint disabing this join type. */
+    private final HintDefinition knownDisableHint;
+
+    /** Hint forcing usage of this join type. */
+    private final HintDefinition knownForceHint;
+
+    static {
+        HINTS.put(NL_JOIN, NO_NL_JOIN);
+        HINTS.put(CNL_JOIN, NO_CNL_JOIN);
+        HINTS.put(MERGE_JOIN, NO_MERGE_JOIN);
+    }
+
+    /** */
+    protected AbstractIgniteJoinConverterRule(String descriptionPrefix, HintDefinition forceHint,
+        HintDefinition disableHint) {
+        super(LogicalJoin.class, descriptionPrefix);
+
+        assert forceHint != disableHint;
+        assert HINTS.containsKey(forceHint);
+        assert HINTS.containsValue(disableHint);
+        assert !HINTS.containsKey(disableHint);
+        assert !HINTS.containsValue(forceHint);
+
+        knownDisableHint = disableHint;
+        knownForceHint = forceHint;
+    }
+
+    /** {@inheritDoc} */
+    @Override public final boolean matches(RelOptRuleCall call) {
+        return super.matches(call) && matchesCall(call) && !disabledByHints(call.rel(0));
+    }
+
+    /** */
+    private boolean disabledByHints(LogicalJoin join) {
+        if (HintUtils.allRelHints(join).isEmpty())
+            return false;
+
+        boolean ruleDisabled = false;
+
+        Map<String, Collection<HintDefinition>> hintedTables = new HashMap<>();
+
+        Set<String> joinTbls = joinTblNames(join);
+
+        assert joinTbls.size() < 3;
+
+        for (RelHint hint : HintUtils.hints(join, HINTS.keySet(), HINTS.values())) {
+            Set<String> matchedTbls = hint.listOptions.isEmpty() ? joinTbls : new HashSet<>(hint.listOptions);
+
+            if (!hint.listOptions.isEmpty())
+                matchedTbls.retainAll(joinTbls);
+
+            if (matchedTbls.isEmpty())
+                continue;
+
+            HintDefinition curHintDef = HintDefinition.valueOf(hint.hintName);
+            boolean curHintIsDisable = !HINTS.containsKey(curHintDef);
+            boolean unableToProcess = false;
+
+            for (String tbl : joinTbls) {
+                Collection<HintDefinition> prevTblHints = hintedTables.get(tbl);
+
+                if (prevTblHints == null)
+                    continue;
+
+                int disableCnt = 0;
+
+                for (HintDefinition prevTblHint : prevTblHints) {
+                    boolean prevHintIsDisable = !HINTS.containsKey(prevTblHint);
+
+                    if (prevHintIsDisable)
+                        ++disableCnt;
+
+                    boolean opposite = prevTblHint != curHintDef && (!prevHintIsDisable || !curHintIsDisable)
+                        && (prevTblHint == HINTS.get(curHintDef) || curHintDef == HINTS.get(prevTblHint)
+                        || (!curHintIsDisable && HINTS.get(prevTblHint) != null));
```

Comment:
Looks overcomplicated. At least `HINTS.get(prevTblHint) != null` can be simplified to `!prevHintIsDisable', perhaps something else.